File: data-structures/stacks/MyStackLinkedList.cpp
// C++ implementation of stack using linked list
#include <iostream>
using namespace std;

struct Node {
	int data;
	Node *next;
	Node *prev;
};

typedef Node *NodePtr;

class MyStackLinkedList {
private:
	NodePtr tail;
public:
	MyStackLinkedList() {
		tail = nullptr;
	}

	// push operation
	void push(int item) {
		NodePtr node = new Node;
		node->data = item;
		node->next = nullptr;
		node->prev = nullptr;

		if (tail == nullptr) {
			tail = node;
		} else {
			tail->next = node;
			node->prev = tail;
			tail = node;
		}
	}
    
    // pop operation
	int pop() {
		if (tail == nullptr) {
			cout<<"ERROR: stack is empty"<<endl;
			return -999999;
		}

		int item = tail->data;
		NodePtr prev = tail->prev;
		if (prev != nullptr) {
			prev->next = nullptr;
		}
		tail->prev = nullptr;
		delete tail;
		tail = prev;
		return item;
	}
	

	// peek (top) operation
	int peek() {
		if (tail == nullptr) {
			cout<<"ERROR: stack is empty"<<endl;
			return -999999;
		}

		return tail->data;
	}

};

int main() {
    MyStackLinkedList stack;
    stack.push(44);
    stack.push(33);
    cout<<stack.pop()<<endl;
    cout<<stack.pop()<<endl;
    cout<<stack.pop()<<endl;
    stack.push(45);
    cout<<stack.peek()<<endl;
	return 0;
}

File: data-structures/stacks/MyStackLinkedList.java
// Java implementation of stack using linked list

class Node {
	int data;
	Node prev;
	Node next;

	public Node(int item) {
		data = item;
		prev = null;
		next = null;
	}
}

public class MyStackLinkedList {
	private Node tail;

	public MyStackLinkedList() {
		tail = null;
	}

	public void push(int item) {
		Node node = new Node(item);

		if (tail == null) {
			tail = node; 
		} else {
			node.prev = tail;
			tail.next = node;
			tail = node;
		}
	}

	public int pop() {
		if (tail == null) {
			System.out.println("ERROR: Stack is empty");
			return -999999;
		}
		int item = tail.data;
		Node prev = tail.prev;
		if (prev != null) {
			prev.next = null;
		}

		tail.prev = null;
		tail = prev;
		return item;
	} 

	public int peek() {
		if (tail == null) {
			System.out.println("ERROR: Stack is empty");
			return -999999;
		}

		return tail.data;
	}

	public static void main(String [] args) {
		MyStackLinkedList stack = new MyStackLinkedList();
		stack.push(45);
		stack.push(66);
		System.out.println(stack.pop());
		System.out.println(stack.peek());
		System.out.println(stack.pop());
		System.out.println(stack.pop());
	}
}
